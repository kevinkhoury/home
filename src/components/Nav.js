/* eslint react/jsx-no-target-blank: 0 */ // --> OFF

import React, { Fragment } from 'react'
import { Link } from 'gatsby'
import PropTypes from 'prop-types'
import { v4 as uuidv4 } from 'uuid'
import { useNavPages, useSiteData } from '../hooks'

const NavLink = ({ slug, name }) => {
  const linkClass = 'nav-' + name.replace(' ', '-').toLowerCase()
  return (
    <li className={linkClass}>
      <Link to={slug}>{name}</Link>
    </li>
  )
}

NavLink.propTypes = {
  slug: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
}

const SocialLink = ({ slug, name }) => (
  <a className="social-link" rel="noopener" target="_blank" href={slug}>
    {name}
  </a>
)

SocialLink.propTypes = {
  slug: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
}

const Nav = ({ toggleNav, setToggleNav, siteName }) => {
  const navPages = useNavPages()
  const {
    socialLinks: { twitter, facebook, linkedin, pinterest, instagram },
  } = useSiteData()
  const socialLinks = [
    // get any social sites that we have
    { Twitter: twitter },
    { Facebook: facebook },
    { LinkedIn: linkedin },
    { Pinterest: pinterest },
    { Instagram: instagram },
  ].filter(
    (item) =>
      !!Object.values(item)[0] &&
      !!Object.values(item)[0].url &&
      !!Object.values(item)[0].show,
  )

  return (
    <Fragment>
      <button
        className="nav-burger"
        onClick={() => setToggleNav(!toggleNav)}
        aria-label="Open or close menu"
      >
        <div className="hamburger hamburger--collapse">
          <div className="hamburger-box">
            <div className="hamburger-inner" />
          </div>
        </div>
      </button>
      <nav id="swup" className="site-head-left">
        <ul className="nav">
          {!!navPages &&
            navPages.length > 1 &&
            navPages.map((page) => (
              <NavLink slug={page.slug} name={page.label} key={uuidv4()} />
            ))}
        </ul>
      </nav>
      <div className="site-head-center">
        <Link className="site-head-logo" to={`/`}>
          {siteName}
        </Link>
      </div>
      <div className="site-head-right">
        <div className="social-links">
          {!!socialLinks &&
            !!socialLinks.length &&
            socialLinks.map((social) => (
              <SocialLink
                slug={Object.values(social)[0].url}
                name={Object.keys(social)[0]}
                key={uuidv4()}
              />
            ))}
        </div>
      </div>
    </Fragment>
  )
}

Nav.propTypes = {
  toggleNav: PropTypes.bool.isRequired,
  siteName: PropTypes.string.isRequired,
}

export default Nav
