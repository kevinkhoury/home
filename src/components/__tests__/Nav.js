import React from 'react'
import {
  useSiteData as useSiteDataMock,
  useNavPages as useNavPagesMock,
} from '../../hooks'
import renderer from 'react-test-renderer'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import Nav from '../Nav'

const DEFAULT_PROPS = {
  toggleNav: false,
  setToggleNav: jest.fn(),
  siteName: 'foobar.com',
}

jest.mock('../../hooks', () => ({
  useSiteData: jest.fn(),
  useNavPages: jest.fn(),
}))

function mockSocialLink(show) {
  return {
    url: 'https://foobar.com',
    show: !!show,
  }
}

beforeEach(() => {
  useSiteDataMock.mockImplementation(() => ({
    socialLinks: {
      twitter: mockSocialLink(true),
      facebook: mockSocialLink(true),
      linkedin: mockSocialLink(false),
      pinterest: mockSocialLink(true),
      instagram: mockSocialLink(false),
    },
  }))
  useNavPagesMock.mockImplementation(() => [
    {
      slug: '/foobar1',
      label: 'Foobar1',
    },
    {
      slug: '/foobar2',
      label: 'Foobar2',
    },
    {
      slug: '/foobar3',
      label: 'Foobar3',
    },
  ])
})

afterEach(() => {
  jest.clearAllMocks()
  __rewire_reset_all__()
})

describe('Nav', () => {
  it('snapshot renders properly', () => {
    const tree = renderer.create(<Nav {...DEFAULT_PROPS} />)
    expect(tree.toJSON()).toMatchSnapshot()
  })
  it('renders all nav items', () => {
    render(<Nav {...DEFAULT_PROPS} />)
    const result1 = screen.getByText(/Foobar1/)
    const result2 = screen.getByText(/Foobar1/)
    const result3 = screen.getByText(/Foobar3/)
    expect(result1).toBeTruthy()
    expect(result2).toBeTruthy()
    expect(result3).toBeTruthy()
  })
  it('renders only social nav items marked with show === true', () => {
    render(<Nav {...DEFAULT_PROPS} />)
    const result1 = screen.getByText(/Twitter/)
    const result2 = screen.getByText(/Facebook/)
    const result3 = screen.getByText(/Pinterest/)
    expect(result1).toBeTruthy()
    expect(result2).toBeTruthy()
    expect(result3).toBeTruthy()
  })
  it('toggles nav on button click', () => {
    render(<Nav {...DEFAULT_PROPS} />)
    userEvent.click(screen.getByRole('button'))
    expect(DEFAULT_PROPS.setToggleNav.mock.calls.length).toBe(1)
  })
})

describe('SocialLink', () => {
  it('snapshot renders properly', () => {
    const SocialLink = Nav.__get__('SocialLink')
    const tree = renderer.create(<SocialLink slug="/slug" name="name" />)
    expect(tree.toJSON()).toMatchSnapshot()
  })
})

describe('NavLink', () => {
  it('snapshot renders properly', () => {
    const NavLink = Nav.__get__('NavLink')
    const tree = renderer.create(<NavLink slug="/slug" name="Name" />)
    expect(tree.toJSON()).toMatchSnapshot()
  })
})
