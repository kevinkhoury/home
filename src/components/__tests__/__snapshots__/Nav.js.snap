// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav snapshot renders properly 1`] = `
Array [
  <button
    aria-label="Open or close menu"
    className="nav-burger"
    onClick={[Function]}
  >
    <div
      className="hamburger hamburger--collapse"
    >
      <div
        className="hamburger-box"
      >
        <div
          className="hamburger-inner"
        />
      </div>
    </div>
  </button>,
  <nav
    className="site-head-left"
    id="swup"
  >
    <ul
      className="nav"
    >
      <li
        className="nav-foobar1"
      >
        <a
          href="/foobar1"
        >
          Foobar1
        </a>
      </li>
      <li
        className="nav-foobar2"
      >
        <a
          href="/foobar2"
        >
          Foobar2
        </a>
      </li>
      <li
        className="nav-foobar3"
      >
        <a
          href="/foobar3"
        >
          Foobar3
        </a>
      </li>
    </ul>
  </nav>,
  <div
    className="site-head-center"
  >
    <a
      className="site-head-logo"
      href="/"
    >
      foobar.com
    </a>
  </div>,
  <div
    className="site-head-right"
  >
    <div
      className="social-links"
    >
      <a
        className="social-link"
        href="https://foobar.com"
        rel="noopener"
        target="_blank"
      >
        Twitter
      </a>
      <a
        className="social-link"
        href="https://foobar.com"
        rel="noopener"
        target="_blank"
      >
        Facebook
      </a>
      <a
        className="social-link"
        href="https://foobar.com"
        rel="noopener"
        target="_blank"
      >
        Pinterest
      </a>
    </div>
  </div>,
]
`;

exports[`NavLink snapshot renders properly 1`] = `
<li
  className="nav-name"
>
  <a
    href="/slug"
  >
    Name
  </a>
</li>
`;

exports[`SocialLink snapshot renders properly 1`] = `
<a
  className="social-link"
  href="/slug"
  rel="noopener"
  target="_blank"
>
  name
</a>
`;
